File: frontend/src/types/menu.ts
import type { RecipeListItem } from './recipe'

/** A recipe of a menu, cooked once and eaten over `portions` meals. */
export interface MenuItem {
  id: number
  portions: number     // number of meals covered by this recipe
  position: number
  recipe: RecipeListItem | null
}

/** A full menu, with its recipes. */
export interface Menu {
  id: number
  name: string
  mealsCount: number   // number of meals asked for at creation
  createdAt: string
  updatedAt: string
  items: MenuItem[]
}

/** Lightweight menu shape — used on the menu list page. */
export interface MenuListItem {
  id: number
  name: string
  mealsCount: number
  recipesCount: number
  createdAt: string
}

/** Parameters of an automatic menu generation. */
export interface MenuGenerateParams {
  name: string
  mealsCount: number
}

/** Nutritional values of a meal (kcal and macros in g). */
export interface MealNutrition {
  calories: number
  proteinsG: number
  carbsG: number
  fatsG: number
}

/** Category of main meals: used by generation and counted in the menu's meals (other categories are extras). */
export const MENU_CATEGORY = 'repas'

/** A line of a menu's shopping list (ingredients merged by name and unit). */
export interface ShoppingListItem {
  name: string
  quantity: number | null  // total for the menu's portions, null if not measurable
  unit: string | null
  aisle: string | null     // store aisle from the ingredients' enrichment (Ciqual / Mistral), null if not enriched
  extras: string[]         // quantities that couldn't be added up, as written
  recipes: string[]        // titles of the recipes that need this ingredient
}

File: backend/app/schemas/menu.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.recipe import RecipeListItem

MAX_MEALS_PER_MENU = 28


# ── Output ─────────────────────────────────────────────────────────────────────

class MenuItemOut(BaseModel):
    id: int
    portions: int  # number of meals covered by this recipe
    position: int
    recipe: RecipeListItem | None
    model_config = ConfigDict(from_attributes=True)


class MenuOut(BaseModel):
    id: int
    name: str
    meals_count: int
    created_at: datetime
    updated_at: datetime
    items: list[MenuItemOut]
    model_config = ConfigDict(from_attributes=True)


class MenuListItem(BaseModel):
    id: int
    name: str
    meals_count: int
    recipes_count: int
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


# ── Input ──────────────────────────────────────────────────────────────────────

class MenuGenerateRequest(BaseModel):
    # automatic generation by Ollama
    name: str = Field(min_length=1, max_length=255)
    meals_count: int = Field(ge=1, le=MAX_MEALS_PER_MENU)


class MenuItemCreate(BaseModel):
    # manual add / replacement of a recipe
    recipe_id: int
    portions: int | None = Field(default=None, ge=1)  # defaults to the recipe's own servings


class MenuUpdate(BaseModel):
    # rename a menu
    name: str = Field(min_length=1, max_length=255)


# ── Shopping list ──────────────────────────────────────────────────────────────

class ShoppingListItem(BaseModel):
    name: str
    quantity: float | None  # total scaled to the menu's portions, None if not measurable
    unit: str | None
    aisle: str | None       # store aisle key, None when the ingredients aren't enriched yet
    extras: list[str]       # quantities that couldn't be added up, as written ("quelques gouttes")
    recipes: list[str]      # titles of the recipes that need this ingredient
    model_config = ConfigDict(from_attributes=True)
